calico/src/main/scala/calico/syntax.scala

```diff
@@ -44,6 +44,13 @@ extension [F[_]](component: Resource[F, dom.Node])
       Resource.make(F.delay(root.appendChild(e)))(_ => F.delay(root.removeChild(e))).void
     }
 
+extension [F[_], A](sigA: Signal[F, A])
+  def mapCache[B](f: A => B)(using F: Concurrent[F]): Resource[F, Signal[F, B]] =
+    given Eq[B] = (x, y) => x.asInstanceOf[js.Any] eq y.asInstanceOf[js.Any]
+    sigA.get.flatMap(a => SignallingRef[F].of(f(a))).toResource.flatTap { sigB =>
+      sigA.discrete.map(f).changes.foreach(sigB.set).compile.drain.background
+    }
```

Comment:
The current implementation starts a subscription to `Signal[A]` even if `Signal[B]` currently has no subscribers. With more work, we could optimize even this—so the subscription to `Signal[A]` is active only if `Signal[B]` has at least one active subscriber.

Reply:
I gave it a go. Caused a few more changes than I expected because this now returns a `Resource`.  I still found I was seeing a lot more Signal evaluations than expected. 

I suspect this is because I was comparing to laminar, and with laminar when splitting a Signal of List into child components, I think there is some magic which memoises the previous value to prevent emitting it again if it hasn't changed. https://github.com/raquo/Airstream/blob/f6023481cdd011334575aeb147fdfe7ff64c0f6b/src/main/scala/com/raquo/airstream/split/SplitEventStream.scala#L29

It might be the two projects have a different approach to how Signals are intended to work.

Reply:
@lukestephenson thanks for giving it a try!

> Caused a few more changes than I expected because this now returns a `Resource`.

Yes, this is fundamental: to reduce evaluations, we have to cache the value somewhere, which means we need mutable state, which can only be allocated effectually. We also need a background process to update this cache, which is done in a `Resource` to ensure it is properly cancelled and cleaned up.

---

> and with laminar when splitting a Signal of List into child components, I think there is some magic which memoises the previous value to prevent emitting it again if it hasn't changed.

Yup, Calico implements this exact same magic :)

https://github.com/armanbilge/calico/blob/032a3a7adbe722da67a6b04ce91dea4f3d1bb6bc/calico/src/main/scala/calico/dsl.scala#L368

> I still found I was seeing a lot more Signal evaluations than expected.

Sorry, do you mind expanding more on this? I would very much like to get to the bottom of this. If you don't mind pushing another commit to the demo project you shared with me that would be very helpful. Thanks!

Reply:
ok.  I've updated my demo project, checkout the readme for a bit more context.

One difference I've noticed is that in laminar, when splitting a `Signal[List[MyDomain]]`, it takes care of providing me with a `Signal[MyDomain]` for each list item.  In calico, when working with keyed children, I just get the "key" provided back to me and then it is my responsibility to filter the original signal by that key.  I wonder if I am doing something wrong there.

With laminar, because it is splitting the `List`, it can have awareness of which elements of the list have changed and do some caching of that.  Calico appears to just be caching the final dom nodes that ?

Actually, I think there is just something in all this that I'm struggling with mentally.  I find it a bit hard to reason about the Signals that aren't cached by default.

Reply:
@lukestephenson thanks, once again this is super helpful!

---

> One difference I've noticed is that in laminar, when splitting a `Signal[List[MyDomain]]`, it takes care of providing me with a `Signal[MyDomain]` for each list item. In calico, when working with keyed children, I just get the "key" provided back to me and then it is my responsibility to filter the original signal by that key. I wonder if I am doing something wrong there.

Ah yeah, you are doing nothing wrong here. You're right, the ergonomics of this can be improved.

> With laminar, because it is splitting the `List`, it can have awareness of which elements of the list have changed and do some caching of that. Calico appears to just be caching the final dom nodes that ?

Oh I see what you mean. This is not just an ergonomics problem, there is an actual efficiency loss.

---

I looked at your updated demo (thanks for that!) and I think I understood the problem. I've currently implemented `.mapCache` to use reference equality i.e. `eq`:

https://github.com/armanbilge/calico/blob/0bf8dfcbf566418285b4c0e029c5b5818013959f/calico/src/main/scala/calico/syntax.scala#L49

But, if you are rebuilding your data structure from scratch on every change, then `eq` does not work. Which in your case, is exactly what `rowSignal` is doing: every time the board changes, it is recreating the `(Int, List[Option[Orientation]]])` completely from scratch, thereby bypassing the reference-equality caching in `mapCache` :)

So the fix would be for Calico to use deep equality via `Eq` instead of reference equality with `eq`. I tried this in 6becace2c8a1634b4cb69e88acf63fc999f11ad1 and opened a PR on your demo project showing the improvement.

There is an interesting design choice here: the great thing about `eq` is that it is fast and cheap; with deep equality via `Eq` you can end up traversing the entire data structure (which could be quite large). And if you add a `.mapCache` for every "layer" of your nested data structure, then the overall amount of work could grow quite a bit—but perhaps I'm overthinking it 🤔 

---

> Actually, I think there is just something in all this that I'm struggling with mentally. I find it a bit hard to reason about the Signals that aren't cached by default.

Yes, I totally get that! Outside of Calico and frontend, how familiar/comfortable are you with Cats Effect, `IO`, and programs-as-values generally?

For example, consider the following program.

```scala
val myIO = IO(42)
val myMappedIO = myIO.map { n =>
  println("doubling")
  2 * n
}

myMappedIO.flatMap(IO.println(_)) *>
  myMappedIO.flatMap(IO.println(_))
```

How many times would you expect the "doubling" `2 * n` to be executed? `Mapping a `Signal` works in exactly the same way.

That is to say: there is no memoization or caching when you `.map` something, because to do so requires state, which can only be created effectually.

Indeed, the memoized version of that example looks like this:
```scala
myMappedIO.memoize.flatMap { memo =>
  memo.flatMap(IO.println(_)) *>
    memo.flatMap(IO.println(_))
}
```

`memoize` is akin to the `mapCache` I've introduced here. In fact, perhaps I should separate it from the `map` and just call it `cache`. 🤔 

---

Thanks again for all the feedback and the demo project, this has been extremely helpful!

Reply:
Thanks for the update.  Outside of calico, I'm very comfortable with programs as values.  Have been working this way for years. It is one of the reasons I was attracted to calico.

On the deep equality vs instance equality, I had the following thoughts:
> But, if you are rebuilding your data structure from scratch on every change,

Ok. So I could switch to a `Map` of rows indexed by row id and only update the affected rows for better performance as well.  It might not be worth optimising this as yet, given my use case is pretty contrived.